compiler/rustc_errors/src/markdown/parse.rs

```diff
@@ -100,25 +96,26 @@ fn parse_recursive<'a>(buf: &'a [u8], ctx: Context) -> MdStream<'_> {
         };
 
         let res: ParseResult<'_> = match (top_blk, prev) {
-            (_, Newline | Whitespace) if loop_buf.starts_with(CMT_S) => {
+            _ if loop_buf.starts_with(CMT_S) => {
                 parse_simple_pat(loop_buf, CMT_S, CMT_E, Po::TrimNoEsc, MdTree::Comment)
             }
             (true, Newline) if loop_buf.starts_with(CBK) => Some(parse_codeblock(loop_buf)),
-            (_, Newline | Whitespace) if loop_buf.starts_with(CIL) => parse_codeinline(loop_buf),
+            _ if loop_buf.starts_with(CIL) => parse_codeinline(loop_buf),
             (true, Newline | Whitespace) if loop_buf.starts_with(HDG) => parse_heading(loop_buf),
             (true, Newline) if loop_buf.starts_with(BRK) => {
                 Some((MdTree::HorizontalRule, parse_to_newline(loop_buf).1))
             }
-            (_, Newline | Whitespace) if loop_buf.starts_with(EMP) => {
-                parse_simple_pat(loop_buf, EMP, EMP, Po::None, MdTree::Emphasis)
+            (_, Newline) if unordered_list_start(loop_buf) => Some(parse_unordered_li(loop_buf)),
+            _ if let Some(strong @ (b"**" | b"__")) = loop_buf.get(..2) => {
+                parse_simple_pat(loop_buf, strong, strong, Po::None, MdTree::Strong)
             }
-            (_, Newline | Whitespace) if loop_buf.starts_with(STG) => {
-                parse_simple_pat(loop_buf, STG, STG, Po::None, MdTree::Strong)
+            _ if let Some(emph @ (b"*" | b"_")) = loop_buf.get(..1) => {
+                parse_simple_pat(loop_buf, emph, emph, Po::None, MdTree::Emphasis)
             }
-            (_, Newline | Whitespace) if loop_buf.starts_with(STK) => {
+            _ if loop_buf.starts_with(STK) => {
                 parse_simple_pat(loop_buf, STK, STK, Po::None, MdTree::Strikethrough)
             }
-            (_, Newline | Whitespace) if loop_buf.starts_with(ANC_S) => {
+            _ if loop_buf.starts_with(ANC_S) => {
```

Comment:
The comment, code, and `*`/`**`/`~~` changes look correct, but `_`/`__` and `<...>` anchors need the preceding whitespace so snake case doesn't turn into italics (gh agrees with commonmark here):

- `foo*bar*`  foo*bar*
- `foo**bar**`  foo**bar**
- `foo_bar_`  foo_bar_
- `foo__bar__` foo__bar__
- `foo~~bar~`  foo~~bar~~
- `` foo`bar` `` foo`bar`
- `foo<www.docs.rs>` foo<www.docs.rs>

Reply:
My impression was it's going for concise over correct hence the hand rolled markdown parser, there's a bunch of other stuff like `(_foo_)` should italicise but not ` _foo _bar` or `foo* bar*`. As far as I can tell these do not occur in the error code docs though, so the question is how much do we want to support?

Reply:
Yeah, it was intended to be pretty simple and just hopefully never receive bad markdown 😆.

Adding `Boundary` to `Prev` and then setting it with something like `x.is_ascii_punctuation() && ~[b'_', b'*', b'~].contains(&x)` might be slightly more accurate to match `Newline | Whitespace | Boundary`. No need for you to do that here though, I'll probably play around with it a bit in a followup.

Reply:
A punctuation boundary sounds good, should I still make the https://github.com/rust-lang/rust/pull/126994#discussion_r1656992434 change or leave it to you when you implement that?

If it ever will support nested formatting then just `x.is_ascii_punctuation()` is closer I think, since `___foo__ bar_` is valid as ___foo__ bar_

Reply:
Up to you, I'll follow up if you prefer not to make the chance here :)

Reply:
Made the change since you would've had to unmerge the `*`/`_`  arms either way